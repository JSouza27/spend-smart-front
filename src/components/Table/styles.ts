import styled, { css } from 'styled-components';

export const Wrapper = styled.div`
  ${({ theme }) => css`
    border: 1px solid ${theme.colors.neutral_300};
    border-radius: ${theme.border.radius};
    width: 100%;

    .rdt_TableHead {
      color: ${theme.colors.neutral_800};
      font-size: ${theme.font.sizes.xxsmall};
      font-weight: ${theme.font.semi_bold};
      line-height: ${theme.spacings.xsmall};
      text-transform: uppercase;
    }

    .rdt_TableHeadRow {
      background-color: ${theme.colors.neutral_50};
    }

    .rdt_TableRow {
      color: ${theme.colors.neutral_900};
      font-size: ${theme.font.sizes.xsmall};
      min-height: 6.4rem;
      line-height: ${theme.spacings.xsmall};
    }
  `}
`;
