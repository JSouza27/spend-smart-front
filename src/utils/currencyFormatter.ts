export class CurrencyFormatter {
  private static language = 'pt-BR';
  private static currency = 'BRL';

  static formatter(num: number): string {
    return (num ?? 0).toLocaleString(this.language, {
      currency: this.currency
    });
  }

  static unformat(num: string): number {
    const nums = num.replace(/[^0-9]+/g, '');
    const i = nums.length - 2;

    return Number(`${nums.substring(0, i)}.${nums.substring(i)}`);
  }
}
