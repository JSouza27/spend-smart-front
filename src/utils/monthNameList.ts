export const monthNameList = [
  'janeiro',
  'fevereiro',
  'mar√ßo',
  'abril',
  'maio',
  'junho',
  'julho',
  'agosto',
  'setembro',
  'outubro',
  'novembro',
  'dezembro'
];
