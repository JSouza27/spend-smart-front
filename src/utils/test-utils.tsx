import { render } from '@testing-library/react';
import React from 'react';
import { ThemeProvider } from 'styled-components';

import theme from '../styles/theme';
import { FormProvider, useForm } from 'react-hook-form';

const AllTheProviders = ({ children }: any) => {
  const methods = useForm();

  return (
    <ThemeProvider theme={theme}>
      <FormProvider {...methods}>{children}</FormProvider>
    </ThemeProvider>
  );
};

const customRender = (
  ui: React.ReactElement<any, string | React.JSXElementConstructor<any>>,
  options?: any
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';
export { customRender as render };
