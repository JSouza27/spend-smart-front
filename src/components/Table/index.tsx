import DataTable from 'react-data-table-component';

import * as S from './styles';

export type TableProps = {
  columns: any;
  data: any;
};

export default function Table({ columns, data }: TableProps) {
  return (
    <S.Wrapper>
      <DataTable
        columns={columns}
        data={data}
        defaultSortFieldId={2}
        pagination
        responsive
      />
    </S.Wrapper>
  );
}
