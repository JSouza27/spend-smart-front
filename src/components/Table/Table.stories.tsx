import { ComponentMeta, ComponentStory } from '@storybook/react';
import { ThemeProvider } from 'styled-components';

import Table from '.';
import theme from '../../styles/theme';
import { CurrencyFormatter } from '../../utils/currencyFormatter';

const columns = [
  { name: 'Descrição', selector: (row: any) => row.description },
  { name: 'Data de Vencimento', selector: (row: any) => row.invoiceDueDate },
  { name: 'Valor', selector: (row: any) => row.value }
];

const row = [
  {
    description: 'Conta de Luz',
    invoiceDueDate: '03/05/2023',
    value: CurrencyFormatter.formatter(255.7),
    type: 'expense',
    sortable: true
  }
];

export default {
  title: 'Table',
  component: Table,
  args: {
    columns,
    data: row
  }
} as ComponentMeta<typeof Table>;

export const Basic: ComponentStory<typeof Table> = (args) => (
  <ThemeProvider theme={theme}>
    <Table {...args} />
  </ThemeProvider>
);
