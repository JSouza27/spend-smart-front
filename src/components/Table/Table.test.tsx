import { render, screen } from 'utils/test-utils';

import Table from '.';
import { CurrencyFormatter } from '../../utils/currencyFormatter';

const columns = [
  { name: 'Descrição', selector: (row: any) => row.description },
  { name: 'Data de Vencimento', selector: (row: any) => row.invoiceDueDate },
  { name: 'Valor', selector: (row: any) => row.value }
];

const row = [
  {
    description: 'Conta de Luz',
    invoiceDueDate: '03/05/2023',
    value: CurrencyFormatter.formatter(255.7),
    type: 'expense',
    sortable: true
  }
];

describe('<Table />', () => {
  it('should render the Table', () => {
    const { container } = render(<Table columns={columns} data={row} />);

    const table = screen.getByRole('table');

    expect(table).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the header', () => {
    render(<Table columns={columns} data={row} />);

    const description = screen.getByText(/descrição/i);
    const date = screen.getByText(/data de vencimento/i);
    const value = screen.getByText(/valor/i);

    expect(description).toBeInTheDocument();
    expect(date).toBeInTheDocument();
    expect(value).toBeInTheDocument();
  });

  it('should render the line', () => {
    render(<Table columns={columns} data={row} />);

    const description = screen.getByText(/conta de luz/i);
    const date = screen.getByText(/03\/05\/2023/i);
    const value = screen.getByText(/r\$ 255,70/i);

    expect(description).toBeInTheDocument();
    expect(date).toBeInTheDocument();
    expect(value).toBeInTheDocument();
  });
});
