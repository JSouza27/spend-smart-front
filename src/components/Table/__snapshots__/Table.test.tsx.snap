// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> should render the Table 1`] = `
.c3 {
  position: relative;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  height: 100%;
  max-width: 100%;
  color: rgba(0,0,0,0.87);
  background-color: #FFFFFF;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  color: rgba(0,0,0,0.87);
  font-size: 12px;
  font-weight: 500;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  width: 100%;
  background-color: #FFFFFF;
  min-height: 52px;
  border-bottom-width: 1px;
  border-bottom-color: rgba(0,0,0,.12);
  border-bottom-style: solid;
}

.c6 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  line-height: normal;
  padding-left: 16px;
  padding-right: 16px;
}

.c12 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  line-height: normal;
  padding-left: 16px;
  padding-right: 16px;
  word-break: break-word;
}

.c7 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  max-width: 100%;
  min-width: 100px;
}

.c13 div:first-child {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  width: 100%;
  box-sizing: border-box;
  font-size: 13px;
  font-weight: 400;
  color: rgba(0,0,0,0.87);
  background-color: #FFFFFF;
  min-height: 48px;
}

.c11:not(:last-of-type) {
  border-bottom-style: solid;
  border-bottom-width: 1px;
  border-bottom-color: rgba(0,0,0,.12);
}

.c8 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: inherit;
  -webkit-justify-content: inherit;
  -ms-flex-pack: inherit;
  justify-content: inherit;
  height: 100%;
  width: 100%;
  outline: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  overflow: hidden;
}

.c9 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  position: relative;
  width: 100%;
  border-radius: inherit;
  overflow-x: auto;
  overflow-y: hidden;
  min-height: 0;
}

.c2 {
  position: relative;
  width: 100%;
  display: table;
}

.c18 {
  cursor: pointer;
  height: 24px;
  max-width: 100%;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  padding-left: 8px;
  padding-right: 24px;
  box-sizing: content-box;
  font-size: inherit;
  color: inherit;
  border: none;
  background-color: transparent;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  direction: ltr;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c18::-ms-expand {
  display: none;
}

.c18:disabled::-ms-expand {
  background: #f60;
}

.c18 option {
  color: initial;
}

.c17 {
  position: relative;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-size: inherit;
  color: inherit;
  margin-top: 1px;
}

.c17 svg {
  top: 0;
  right: 0;
  color: inherit;
  position: absolute;
  fill: currentColor;
  width: 24px;
  height: 24px;
  display: inline-block;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  pointer-events: none;
}

.c14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  padding-right: 8px;
  padding-left: 8px;
  width: 100%;
  color: rgba(0,0,0,0.54);
  font-size: 13px;
  min-height: 56px;
  background-color: #FFFFFF;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: rgba(0,0,0,.12);
}

.c21 {
  position: relative;
  display: block;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border: none;
  border-radius: 50%;
  height: 40px;
  width: 40px;
  padding: 8px;
  margin: px;
  cursor: pointer;
  -webkit-transition: 0.4s;
  transition: 0.4s;
  color: rgba(0,0,0,.54);
  fill: rgba(0,0,0,.54);
  background-color: transparent;
}

.c21:disabled {
  cursor: unset;
  color: rgba(0,0,0,.18);
  fill: rgba(0,0,0,.18);
}

.c21:hover:not(:disabled) {
  background-color: rgba(0,0,0,.12);
}

.c21:focus {
  outline: none;
  background-color: rgba(0,0,0,.12);
}

.c20 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 4px;
  white-space: nowrap;
}

.c15 {
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c19 {
  margin: 0 24px;
}

.c16 {
  margin: 0 4px;
}

.c0 {
  border: 1px solid #E6E8F0;
  border-radius: 0.4rem;
  width: 100%;
}

.c0 .rdt_TableHead {
  color: #474D66;
  font-size: 1.0rem;
  font-weight: 600;
  line-height: 1.6rem;
  text-transform: uppercase;
}

.c0 .rdt_TableHeadRow {
  background-color: #FAFBFF;
}

.c0 .rdt_TableRow {
  color: #101840;
  font-size: 1.2rem;
  min-height: 6.4rem;
  line-height: 1.6rem;
}

@media screen and (max-width:599px) {
  .c20 {
    width: 100%;
    -webkit-box-pack: space-around;
    -webkit-justify-content: space-around;
    -ms-flex-pack: space-around;
    justify-content: space-around;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3 rdt_Table"
        role="table"
      >
        <div
          class="c4 rdt_TableHead"
          role="rowgroup"
        >
          <div
            class="c5 rdt_TableHeadRow"
            role="row"
          >
            <div
              class="c6 c7 rdt_TableCol"
              data-column-id="1"
            >
              <div
                class="c8 rdt_TableCol_Sortable"
                data-column-id="1"
                data-sort-id="1"
                disabled=""
                role="columnheader"
                tabindex="0"
              >
                <div
                  class="c9"
                  data-column-id="1"
                >
                  Descrição
                </div>
              </div>
            </div>
            <div
              class="c6 c7 rdt_TableCol"
              data-column-id="2"
            >
              <div
                class="c8 rdt_TableCol_Sortable"
                data-column-id="2"
                data-sort-id="2"
                disabled=""
                role="columnheader"
                tabindex="0"
              >
                <div
                  class="c9"
                  data-column-id="2"
                >
                  Data de Vencimento
                </div>
              </div>
            </div>
            <div
              class="c6 c7 rdt_TableCol"
              data-column-id="3"
            >
              <div
                class="c8 rdt_TableCol_Sortable"
                data-column-id="3"
                data-sort-id="3"
                disabled=""
                role="columnheader"
                tabindex="0"
              >
                <div
                  class="c9"
                  data-column-id="3"
                >
                  Valor
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="c10 rdt_TableBody"
          role="rowgroup"
        >
          <div
            class="c11 rdt_TableRow"
            id="row-0"
            role="row"
          >
            <div
              class="c12 c7 c13 rdt_TableCell"
              data-column-id="1"
              data-tag="allowRowEvents"
              id="cell-1-undefined"
              role="cell"
            >
              <div
                data-tag="allowRowEvents"
              >
                Conta de Luz
              </div>
            </div>
            <div
              class="c12 c7 c13 rdt_TableCell"
              data-column-id="2"
              data-tag="allowRowEvents"
              id="cell-2-undefined"
              role="cell"
            >
              <div
                data-tag="allowRowEvents"
              >
                03/05/2023
              </div>
            </div>
            <div
              class="c12 c7 c13 rdt_TableCell"
              data-column-id="3"
              data-tag="allowRowEvents"
              id="cell-3-undefined"
              role="cell"
            >
              <div
                data-tag="allowRowEvents"
              >
                R$ 255,70
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div>
    <nav
      class="c14 rdt_Pagination"
    >
      <span
        class="c15 c16"
      >
        Rows per page:
      </span>
      <div
        class="c17"
      >
        <select
          aria-label="Rows per page:"
          class="c18"
        >
          <option
            selected=""
            value="10"
          >
            10
          </option>
          <option
            value="15"
          >
            15
          </option>
          <option
            value="20"
          >
            20
          </option>
          <option
            value="25"
          >
            25
          </option>
          <option
            value="30"
          >
            30
          </option>
        </select>
        <svg
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
        </svg>
      </div>
      <span
        class="c15 c19"
      >
        1-1 of 1
      </span>
      <div
        class="c20"
      >
        <button
          aria-disabled="true"
          aria-label="First Page"
          class="c21"
          disabled=""
          id="pagination-first-page"
          type="button"
        >
          <svg
            aria-hidden="true"
            height="24"
            role="presentation"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M18.41 16.59L13.82 12l4.59-4.59L17 6l-6 6 6 6zM6 6h2v12H6z"
            />
            <path
              d="M24 24H0V0h24v24z"
              fill="none"
            />
          </svg>
        </button>
        <button
          aria-disabled="true"
          aria-label="Previous Page"
          class="c21"
          disabled=""
          id="pagination-previous-page"
          type="button"
        >
          <svg
            aria-hidden="true"
            height="24"
            role="presentation"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
        </button>
        <button
          aria-disabled="true"
          aria-label="Next Page"
          class="c21"
          disabled=""
          id="pagination-next-page"
          type="button"
        >
          <svg
            aria-hidden="true"
            height="24"
            role="presentation"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
        </button>
        <button
          aria-disabled="true"
          aria-label="Last Page"
          class="c21"
          disabled=""
          id="pagination-last-page"
          type="button"
        >
          <svg
            aria-hidden="true"
            height="24"
            role="presentation"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.59 7.41L10.18 12l-4.59 4.59L7 18l6-6-6-6zM16 6h2v12h-2z"
            />
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
            />
          </svg>
        </button>
      </div>
    </nav>
  </div>
</div>
`;
